### PR TITLE
Constify parameters and return values of core APIs

### PR DESCRIPTION
**Pull request by erimatnor**
Harden core APIs by adding the `const` qualifier to pointer parameters
and return values passed by reference. Adding `const` to APIs has
several benefits and potentially reduces bugs.

* Allows core APIs to be called using `const` objects.
* Callers know that objects passed by reference are not modified as a
  side-effect of a function call.
* Returning `const` pointers enforces "read-only" usage of pointers to
  internal objects, forcing users to copy objects when mutating them
  or using explicit APIs for mutations.
* Allows compiler to apply optimizations and helps static analysis.

Note that these changes are so far only applied to core API
functions. Further work can be done to improve other parts of the
code.

### CODE REVIEW

File: tsl/src/chunk_api.c [hunk 1]
```diff
@@ -485,8 +483,8 @@ chunk_api_create_on_data_nodes(Chunk *chunk, Hypertable *ht)
 			DatumGetCString(values[AttrNumberGetAttrOffset(Anum_create_chunk_schema_name)]);
 		table_name = DatumGetCString(values[AttrNumberGetAttrOffset(Anum_create_chunk_table_name)]);
 
-		if (namestrcmp(&chunk->fd.schema_name, schema_name) != 0 ||
-			namestrcmp(&chunk->fd.table_name, table_name) != 0)
+		if (namestrcmp((Name) &chunk->fd.schema_name, schema_name) != 0 ||
+			namestrcmp((Name) &chunk->fd.table_name, table_name) != 0)
```
Comment (mkindahl) [inline tsl/src/chunk_api.c:487]: Not sure why these casts are here. Neither the const-ness of `schema_name` nor the `table_name` has changed, so why are these necessary?

**Review by mkindahl — COMMENTED** [review]
I wonder about mainly two things:

- There are copying added to several objects, but changing to using `const` does not affect the lifetime of objects so I wonder why these are needed.
- There are casts added to cast away const-ness when assigning to fields but it seems more appropriate to add `const` to the fields. For example, it seems like `CollisionInfo::cube` could be a pointer to a constant `Hypercube`.

File: src/dimension.c [hunk 2]
```diff
@@ -1619,13 +1631,12 @@ dimension_rename_schema_name(TupleInfo *ti, void *data)
...
 void
-ts_dimensions_rename_schema_name(char *old_name, char *new_name)
+ts_dimensions_rename_schema_name(const char *old_name, const char *new_name)
 {
 	NameData old_schema_name;
 	ScanKeyData scankey[1];
 	Catalog *catalog = ts_catalog_get();
-	char *names[2] = { old_name, new_name };
-
+	char *names[2] = { pstrdup(old_name), pstrdup(new_name) };
```
Comment (svenklemm) [inline src/dimension.c:1639]: why do we need to copy these now?

File: src/trigger.c [hunk 3]
```diff
@@ -53,8 +54,8 @@ ts_trigger_create_on_chunk(Oid trigger_oid, char *chunk_schema_name, char *chunk
 	} while (0);
 
 	Assert(IsA(stmt, CreateTrigStmt));
-	stmt->relation->relname = chunk_table_name;
-	stmt->relation->schemaname = chunk_schema_name;
+	stmt->relation->relname = pstrdup(chunk_table_name);
+	stmt->relation->schemaname = pstrdup(chunk_schema_name);
```
Comment (svenklemm) [inline src/trigger.c:58]: why do we need the copy now?

File: src/chunk.c [hunk 4]
```diff
@@ -491,11 +492,11 @@ check_for_collisions(ChunkScanCtx *scanctx, ChunkStub *stub)
  * chunk.
  */
 static ChunkStub *
-chunk_collides(Hypertable *ht, Hypercube *hc)
+chunk_collides(const Hypertable *ht, const Hypercube *hc)
 {
 	ChunkScanCtx scanctx;
 	CollisionInfo info = {
-		.cube = hc,
+		.cube = (Hypercube *) hc,
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:499]: I think I tried this and it does not work due to cascading to other things, but will check again.

File: src/chunk.c [hunk 5]
```diff
@@ -774,23 +775,24 @@ copy_hypertable_acl_to_relid(Hypertable *ht, Oid relid)
...
 	 * The CreateForeignTableStmt embeds a regular CreateStmt, so we can use
 	 * it to create both regular and foreign tables
 	 */
 	CreateForeignTableStmt stmt = {
 		.base.type = T_CreateStmt,
-		.base.relation =
-			makeRangeVar(NameStr(chunk->fd.schema_name), NameStr(chunk->fd.table_name), 0),
-		.base.inhRelations = list_make1(makeRangeVar(namespace, hyper_name, 0)),
+		.base.relation = makeRangeVar(pstrdup(NameStr(chunk->fd.schema_name)),
+									  pstrdup(NameStr(chunk->fd.table_name)),
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:793]: The chunk is now `const` so we cannot reference a `const` string in `makeRangeVar` since PostgreSQL expects mutable strings.

In general, PostgreSQL often mutates planner nodes, so to be safe we shouldn't pass in pointers to strings that also exist in other objects (like the chunk in this case). The fact that the chunk is `const` here actually makes it obvious that we shouldn't pass it's name and schema strings to the planner node without copying them first. This is a good thing, IMO, and might prevent a bunch of subtle bugs.

File: src/chunk.c [hunk 6]
```diff
@@ -774,23 +775,24 @@ copy_hypertable_acl_to_relid(Hypertable *ht, Oid relid)
...
+									  0),
+		.base.inhRelations = list_make1(makeRangeVar(pstrdup(namespace), pstrdup(hyper_name), 0)),
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:795]: Same answer as above.

File: src/hypertable.c [hunk 7]
```diff
@@ -118,18 +118,18 @@ ts_hypertable_permissions_check_by_id(int32 hypertable_id)
...
 
 static Oid
-get_chunk_sizing_func_oid(FormData_hypertable *fd)
+get_chunk_sizing_func_oid(const FormData_hypertable *fd)
 {
 	Oid argtype[] = { INT4OID, INT8OID, INT8OID };
-	return LookupFuncName(list_make2(makeString(NameStr(fd->chunk_sizing_func_schema)),
-									 makeString(NameStr(fd->chunk_sizing_func_name))),
+	return LookupFuncName(list_make2(makeString(pstrdup(NameStr(fd->chunk_sizing_func_schema))),
+									 makeString(pstrdup(NameStr(fd->chunk_sizing_func_name)))),
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/hypertable.c:125]: `makeString` expects a mutable string, so there is no guarantee that it won't change the string contents. Since the hypertable metadata is now `const` we need to make a copy of the string so that we know the original string is not modified by other code.

File: tsl/src/chunk_api.c (hunk 1 above)
Comment (erimatnor) [reply tsl/src/chunk_api.c:487]: Check the signature of `namestrcmp`. For some reason, `namestrcmp` takes a mutable (non-const) `Name` and a `const` string (`const char *`) and compares them. Normally, `strcmp` takes two `const` pointers, but not in PostgreSQL apparently. I cast it since I know that `namestrcmp` does not mutate the `Name` argument.

File: src/dimension.c (hunk 2 above)
Comment (erimatnor) [reply src/dimension.c:1639]: because `old_name` and `new_name` are now `const char *` pointers but scanning function expects a non-const data argument. I could cast away the `const`, but it seems to me that this is safer and prevents bugs in case of future changes.

File: src/trigger.c (hunk 3 above)
Comment (erimatnor) [reply src/trigger.c:58]: PostgresSQL more-or-less always expects mutable objects/strings in planner nodes. The trigger function called here now takes non-mutable strings parameters (`chunk_table_name` and `chunk_schema_name`), so copies are needed to guarantee that the original strings are not modified.

IMO, this is a good thing as it is much safer and prevents a number of subtle bugs.

File: src/chunk.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:793]: But adding `const` does not in itself make it necessary to copy data, and the implementation of `makeRangeVar` does not mutate the string. It is only non-`const` because the PG interface is not `const`-correct.

The only reasons I can see for copying the string is that:
- The source has a shorter life-time than the destination. Changing from non-`const` to `const` does not protect against this anyway, and since the code already works, this could not be a reason to copy the string.
- The destination need to be modified. This is what adding `const` protects against, but `makeRangeVar` does not modify the string, so there is no reason to copy the string. However, it does not copy the string either, it just assigns it to the `schema_name` field, but looking over the code I do not see any place where the schema name is actually modified or rewritten. Yes, it is quite often assigned a different value, but that does not mean that the original string is modified, which would be a reason for copying the string.

The danger with copying the string is that we suddenly start to allocate more memory for the strings, which we should not start to do as part of a patch that should just refactor the code.

File: src/chunk.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:795]: Same comment as above.

File: src/hypertable.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/hypertable.c:125]: Function `makeString` takes a mutable string because it is not `const`-correct, not because it intend to modify the string (it could just as well accept a `const char*`). The lifetime of the function name list is just for the duration of the lookup, so the copying is not necessary, especially not since this is intended to be a refactoring.

File: tsl/src/chunk_api.c (hunk 1 above)
Comment (mkindahl) [reply tsl/src/chunk_api.c:487]: The type of `chunk->fd.schema_name` is `NameData` and AFAICT you haven't changed that in this commit. Taking the address of `chunk->fd.schema_name` gives a `NameData*`, which is the same as `Name`. This is what `namestrcmp` accepts as the first argument, so that would mean that cast is redundant. What am I missing?

File: src/chunk.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:499]: If it cascades, you can leave it for a later PR.

**Review by mkindahl — COMMENTED** [review]
I think you should avoid to start copying strings in a refactoring patch and keep the scope down to just constifying the code.

File: src/chunk.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:793]: I don't think just looking at the "constructor" of an object, like `makeRangeVar` is enough to conclude that the object or its contained sub-objects won't be modified. You also need to consider the code paths where the object is actually used.

Yes, `makeRangeVar` doesn't mutate the string, but it is commonly used as an object/node in a planner tree, where there is no guarantee it won't be mutated. We often mutate planner trees and objects in our own code, which has also been a source of bugs in the past.

I see no danger with allocating memory for strings. Quite the opposite, I see more dangers with passing around references that you never know where or when they are modified. Heap allocation in PostgreSQL is cheap given the memory context functionality and the memory is freed automatically, so I see little reason to judiciously try to save on memory. The `makeRangeVar` itself is created in the current memory context, so it make sense to copy also the passed in strings so that the range var is a deep copy instead of a shallow one. I guess it takes non-const string pointers so that one can decide on whether a deep or shallow copy makes sense. Clearly, a fully independent object is much safer than passing around references.

**Review by svenklemm — APPROVED** [review]

File: src/dimension.c (hunk 2 above)
Comment (mfundul) [reply src/dimension.c:1639]: Since `ScannerCtx scanctx` is in the stack then `.data = names,` will survive only until the function call returns. In that case shouldn't we explicitly deallocate at the end of the function instead of relying on the current memory context?

File: src/chunk.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:793]: > I am actually not sure what you are proposing. Do you prefer that we cast away the `const` in order to make references in `makeRangeVar` just to save a tiny amount of heap memory?

No, to cast away the `const` and not try to fix a non-existing problem in a refactoring commit.

The lifetime of `stmt` extends to the end of the function. This means that the lifetime of fields of `stmt` also are to the end of the function. Since `chunk` is a parameter, the lifetime of `chunk` extends beyond the function, which means that anything accessed in the lifetime of `stmt` can also be accessed in the lifetime of `chunk`. This means that there is no reason to copy data from `chunk` to `stmt` since data available in `chunk` will exist for the duration of `stmt` (unless memory is deliberately deallocated in the function, which it is not).

[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:793]: @mkindahl These changes does not address any lifecycle/lifetime issues, so not sure why that's a concern. These changes address issues related to mutability, not lifetime. Marking something as `const` means that that object should not be modified. Thus, passing that const object to a function that doesn't take a const version of the object as input implicitly means that the function can modify the passed in object. Therefore, as a rule, it makes sense to pass in a mutable copy and  not a reference to the object, since that is the only way to guarantee that the object is not mutated in the function, nor in a future update to that function.

File: src/hypertable.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/hypertable.c:125]: Again, this PR does not address any lifetime issues. 

We can't on a case-by-case basis decide if a function is "const-correct", then we cannot rely on the interfaces given. We'd have to assume for safety that is is not.

You are also missing that the object returned by `makeString` could itself be passed to a function or be part of a containing object. So, even if `makeString` doesn't itself modify the string object within, another part of the code could. So, it is a false assumption to say that just because a wrapper around one object is not itself mutating the object within the wrapper then no other code can modify it.

File: tsl/src/chunk_api.c (hunk 1 above)
Comment (erimatnor) [reply tsl/src/chunk_api.c:487]: the chunk is `const Chunk *`, which means that also `chunk->fd.schema_name` is `const NameData *`. The function `namestrcmp` doesn't accept this type since it expects a non-const object.

File: src/chunk.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:793]: FWIW, removing the pstrdup leads to this compiler error:

```
/Users/enordstr/Projects/timescaledb/src/chunk.c:796:33: error: passing 'char const[64]' to parameter of type 'char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
                .base.relation = makeRangeVar(NameStr(chunk->fd.schema_name),

```

So, the only other options from making a pstrdup copy is to cast away the `const` or revert the change making `Chunk *` a `const Chunk *`. But the latter is the whole point of this PR: to constify objects passed around in the code.

File: src/dimension.c (hunk 2 above)
Comment (erimatnor) [reply src/dimension.c:1639]: @mfundul it is standard practice in PostgreSQL to defer deallocation of palloc'd objects to the point when the memory context is deleted. This happens all the time. The only thing to be concerned about is if we would allocate unbounded memory, e.g., in a loop over an undefined number of objects.  This is not the case here however.

File: src/chunk.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk.c:505]: I double-checked this. So `CollisionInfo` is used for two types of scans, one where the `.cube` is set during the scan (to find a cube) and another scan where the `.cube` is input data for the scan (which is the case shown here). Therefore, the same pointer is used in both cases.

I guess one alternative option is to have two pointers, one which is the input arg and one mutable, which is the output arg.

File: src/trigger.c (hunk 3 above)
Comment (gayyappan) [reply src/trigger.c:58]: IMO we should not unnecessarily copy objects just to preserve const correctness that is not enforced by PG APIs. This PR introduces multiple unnecessary pstrdups.  It is preferable to either a)  explicitly cast away const i.e. stmt->relation->relname = (char*)chunk_table_name if we want to use const char* arg types in our APIs. or b) Not introduce const char* where it cannot be passed down to the underlying PG function.

**Comment by gayyappan** [comment]
> I wonder about mainly two things:
> 
> * There are copying added to several objects, but changing to using `const` does not affect the lifetime of objects so I wonder why these are needed.
> * There are casts added to cast away const-ness when assigning to fields but it seems more appropriate to add `const` to the fields. For example, it seems like `CollisionInfo::cube` could be a pointer to a constant `Hypercube`.

I agree that there are several places where this PR tries to enforce const correctness by copying strings using pstrdup unnecessarily. This could be addressed by casting away const-ness instead if we want to pass in "const char*" in our APIs , and PG does not respect const in the PG API. 
I also wonder about the correctness of labeling function arguments as const when the operation is actually a mutating operation, such as passing const Chunk to a function that  performs ddl on the chunk, e..g chunk_create_table_constraints, ts_create_trigger_on_chunk, create_default_indexes etc. These might not change the Chunk data structure, but conceptually modify the chunk object in the db.

File: src/trigger.c (hunk 3 above)
Comment (erimatnor) [reply src/trigger.c:58]: FWIW, our code deviates from common PostgreSQL practice by overuse of references instead of safely copying strings and other objects. Here's just a few examples in PostgreSQL:

https://github.com/postgres/postgres/blob/caba8f0d43fb679c6f9643456080408a6bc370e8/src/backend/commands/tablecmds.c#L2618

https://github.com/postgres/postgres/blob/caba8f0d43fb679c6f9643456080408a6bc370e8/src/backend/commands/trigger.c#L1901

https://github.com/postgres/postgres/blob/caba8f0d43fb679c6f9643456080408a6bc370e8/src/backend/foreign/foreign.c#L142

pstrdup is used extensively in PostgreSQL itself, and is cheap given the memory context system. So, I am not sure what we are trying to optimize for by minimizing pstrdups. In fact, we freely heap-allocate and copy other objects in the code and people generally don't mind. Every makeXYZ command and palloc allocates memory in the same way, but for some reason we try hard to not allocate strings even though they are referenced in a palloced object.

**Comment by erimatnor** [comment]
Update: 

* Removed all pstrdup since majority prefers casting away `const` instead of making sure mutable objects avoid references to const objects.

**Comment by erimatnor** [comment]
> > I wonder about mainly two things:
> > 
> > * There are copying added to several objects, but changing to using `const` does not affect the lifetime of objects so I wonder why these are needed.
> > * There are casts added to cast away const-ness when assigning to fields but it seems more appropriate to add `const` to the fields. For example, it seems like `CollisionInfo::cube` could be a pointer to a constant `Hypercube`.
> 
> I agree that there are several places where this PR tries to enforce const correctness by copying strings using pstrdup unnecessarily. This could be addressed by casting away const-ness instead if we want to pass in "const char*" in our APIs , and PG does not respect const in the PG API.
> I also wonder about the correctness of labeling function arguments as const when the operation is actually a mutating operation, such as passing const Chunk to a function that performs ddl on the chunk, e..g chunk_create_table_constraints, ts_create_trigger_on_chunk, create_default_indexes etc. These might not change the Chunk data structure, but conceptually modify the chunk object in the db.

It sounds like you are arguing that functions that have side effects cannot have const objects as input even if those side effects doesn't affect that passed in object?

I personally think that is conflating different issues, but that is for another discussion I think. I changed the code to cast away const instead of copying strings so that I can get the PR approved.

Still, I don't understand why we care about a few pstrdup when we freely allocate (palloc) and copy other objects, literally in the same functions.

File: src/hypertable.c [hunk 8]
```diff
@@ -1368,7 +1369,7 @@ hypertable_check_associated_schema_permissions(const char *schema_name, Oid user
 }
 
 static bool
-relation_has_tuples(Relation rel)
+relation_has_tuples(const Relation rel)
```
Comment (mkindahl) [inline src/hypertable.c:1372]: This expands to `struct RelationData *const`, not `const struct RelationData *`. Not sure if that is what you intended.

**Review by mkindahl — APPROVED** [review]
LGTM

**Review by mfundul — APPROVED** [review]

File: src/hypertable.c [hunk 9]
```diff
@@ -118,18 +118,18 @@ ts_hypertable_permissions_check_by_id(int32 hypertable_id)
...
+	return LookupFuncName(list_make2(makeString((char *) NameStr(fd->chunk_sizing_func_schema)),
+									 makeString((char *) NameStr(fd->chunk_sizing_func_name))),
 						  sizeof(argtype) / sizeof(argtype[0]),
 						  argtype,
 						  false);
 }
 
 static HeapTuple
-hypertable_formdata_make_tuple(const FormData_hypertable *fd, TupleDesc desc)
+hypertable_formdata_make_tuple(const FormData_hypertable *fd, const TupleDesc desc)
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/hypertable.c:132]: I missed that, good catch.

File: src/hypertable.c (hunk 8 above)
Comment (erimatnor) [reply src/hypertable.c:1372]: Missed this one too.

File: src/dimension.c (hunk 2 above)
Comment (erimatnor) [reply src/dimension.c]: This was turned into a cast per request.

**Review by gayyappan — APPROVED** [review]
